### PR TITLE
Managed dependency download on background thread

### PR DESCRIPTION
**Pull request by SatishRanjan**
This PR has the changes to download managed dependency on a background thread and show the managed dependency download in progress message on the azure portal and on the CLI console when a function invocation is triggered while managed dependency download is in-progress

### CODE REVIEW

**Comment by SatishRanjan** [comment]
@daxian-dbw - could you please take a look at this PR

File: src/RequestProcessor.cs [hunk 1]
```diff
@@ -299,27 +321,15 @@ internal StreamingMessage ProcessFunctionEnvironmentReloadRequest(StreamingMessa
...
-                _powershellPool.Initialize(request.RequestId, _dependencyManager.InstallFunctionAppDependencies);
-                response.FunctionLoadResponse.IsDependencyDownloaded = true;
-            }
-            else
+            FunctionLoader.SetupWellKnownPaths(request.FunctionLoadRequest);
+            if (functionLoadRequest.ManagedDependencyEnabled)
             {
-                _powershellPool.Initialize(request.RequestId);
+                //Start dependency download on a separate thread
+                _dependencyDownloadTask = Task.Run(() => _dependencyManager.ProcessDependencies(_msgStream, request)).ContinueWith((task) =>
```
Comment (daxian-dbw) [inline src/RequestProcessor.cs]: The task creation and the reference to the task should be move to `DependencyManager`. For example, `DependencyManager.Initialize(...)` creates the task and hold a reference to it within `DependencyManager`. In `ProcessInvocationRequest` we can wait on the task by `_dependencyManager.DownloadTask.Wait()` or `_dependencyManager.WaitDownloadToFinish()`.

File: src/RequestProcessor.cs [hunk 2]
```diff
@@ -299,27 +321,15 @@ internal StreamingMessage ProcessFunctionEnvironmentReloadRequest(StreamingMessa
...
+                {
+                    _dependencyDownloadTask = null;
```
Comment (daxian-dbw) [inline src/RequestProcessor.cs]: Are you doing this only for the task to be GC'ed?
Doing this in the continuation means it's theoretically possible that `_dependencyDownloadTask` could become null right before calling `_dependencyDownloadTask.Wait();` in `ProcessInvocationRequest`.

After you move the task creation code to `DependencyManager`, you can set the field to null in `_dependencyManager.WaitDownloadToFinish()`, after the waiting on the task returns.

File: src/resources/PowerShellWorkerStrings.resx [hunk 3]
```diff
@@ -26,41 +26,41 @@
...
     mimetype: application/x-microsoft.net.object.bytearray.base64
-    value   : The object must be serialized into a byte array
+    value   : The object must be serialized into a byte array 
             : using a System.ComponentModel.TypeConverter
             : and then encoded with base64 encoding.
     -->
-  <xsd:schema id="root" xmlns="" xmlns:xsd="https://www.w3.org/2001/XMLSchema" xmlns:msdata="urn:schemas-microsoft-com:xml-msdata">
-    <xsd:import namespace="https://www.w3.org/XML/1998/namespace" />
+  <xsd:schema id="root" xmlns="" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:msdata="urn:schemas-microsoft-com:xml-msdata">
+    <xsd:import namespace="http://www.w3.org/XML/1998/namespace" />
```
Comment (daxian-dbw) [inline src/resources/PowerShellWorkerStrings.resx]: Please revert changes in these 2 lines.

File: src/resources/PowerShellWorkerStrings.resx [hunk 4]
```diff
@@ -1,17 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
 <root>
-  <!--
-    Microsoft ResX Schema
-
+  <!--
```
Comment (daxian-dbw) [inline src/resources/PowerShellWorkerStrings.resx]: Please revert all the extra trailing spaces introduced in this file.

File: src/RequestProcessor.cs [hunk 5]
```diff
@@ -209,6 +211,28 @@ internal StreamingMessage ProcessInvocationRequest(StreamingMessage request)
                 functionInfo = _functionLoader.GetFunctionInfo(request.InvocationRequest.FunctionId);
                 psManager = _powershellPool.CheckoutIdleWorker(request, functionInfo);
 
+                if (_dependencyDownloadTask != null
+                    && ((_dependencyDownloadTask.Status != TaskStatus.Canceled)
+                    || _dependencyDownloadTask.Status != TaskStatus.Faulted
+                    || _dependencyDownloadTask.Status != TaskStatus.RanToCompletion))
+                {
+                    psManager.Logger.Log(LogLevel.Information, PowerShellWorkerStrings.DependencyDownloadInProgress, null, true);
+                    psManager.Logger.Log(LogLevel.Information, PowerShellWorkerStrings.DependencyDownloadInProgress, null);
```
Comment (daxian-dbw) [inline src/RequestProcessor.cs]: Why writing this again into the system log? This message is only for user to see as a UX experience improvement, isn't it?

File: src/RequestProcessor.cs [hunk 6]
```diff
@@ -209,6 +211,28 @@ internal StreamingMessage ProcessInvocationRequest(StreamingMessage request)
...
+
+                if (_dependencyManager?.DependencyError != null)
+                {
+                    StreamingMessage response = NewStreamingMessageTemplate(request.RequestId,
+                        StreamingMessage.ContentOneofCase.InvocationResponse,
+                        out StatusResult status);
+                    status.Status = StatusResult.Types.Status.Failure;
+                    status.Exception = _dependencyManager.DependencyError.ToRpcException();
+                    response.InvocationResponse.InvocationId = request.InvocationRequest.InvocationId;
+                    return response;
```
Comment (daxian-dbw) [inline src/RequestProcessor.cs]: As far as I understand, returning a error response to host in `FunctionLoadRequest` phase meaning a hard failure -- all subsequent invocation will fail with the same error.
Failure in `InvocationRequest` phase is not considered a hard failure. This means the next invocation comes in, and it will just continue to execute just like the dependency download/check was successful.

File: src/RequestProcessor.cs [hunk 7]
```diff
@@ -234,7 +257,6 @@ internal StreamingMessage ProcessInvocationRequest(StreamingMessage request)
                 response.InvocationResponse.InvocationId = request.InvocationRequest.InvocationId;
                 status.Status = StatusResult.Types.Status.Failure;
                 status.Exception = e.ToRpcException();
-
```
Comment (daxian-dbw) [inline src/RequestProcessor.cs]: Please add back those two new lines. They make the code more readable.

File: src/DependencyManagement/DependencyManager.cs [hunk 8]
```diff
@@ -17,8 +17,11 @@
 
 namespace Microsoft.Azure.Functions.PowerShellWorker.DependencyManagement
 {
+    using Microsoft.Azure.Functions.PowerShellWorker.Messaging;
     using System.Management.Automation;
     using System.Management.Automation.Language;
+    using System.Management.Automation.Runspaces;
+    using System.Threading;
```
Comment (daxian-dbw) [inline src/DependencyManagement/DependencyManager.cs]: Please move `using Microsoft.Azure.Functions.PowerShellWorker.Messaging;` and `using System.Threading;` to the outer using directives.
The reason `using System.Management.Automation;` is put here is because that's how you can use `PowerShell` type in the following code without having to reference to it by the full type name. So it's OK to put all `System.Management.Automation.*` namespace directive here, but not the rest.

File: src/PowerShell/PowerShellManager.cs [hunk 9]
```diff
@@ -33,6 +33,11 @@ internal class PowerShellManager
         /// </summary>
         internal ILogger Logger => _logger;
 
+        /// <summary>
+        /// Gets the associated PowerShell instance
+        /// </summary>
+        internal PowerShell PowerShellInstance => _pwsh;
```
Comment (daxian-dbw) [inline src/PowerShell/PowerShellManager.cs]: This is not used anywhere, so please remove it.

File: src/PowerShell/PowerShellManagerPool.cs [hunk 10]
```diff
@@ -50,14 +50,14 @@ internal PowerShellManagerPool(MessagingStream msgStream)
         /// Initialize the pool and populate it with PowerShellManager instances.
         /// We instantiate PowerShellManager instances in a lazy way, starting from size 1 and increase the number of workers as needed.
         /// </summary>
-        internal void Initialize(string requestId, Action<PowerShell, ILogger> initAction = null)
+        internal void Initialize(string requestId)
```
Comment (daxian-dbw) [inline src/PowerShell/PowerShellManagerPool.cs]: This method can be removed if not needed anymore.

File: src/PowerShell/PowerShellManagerPool.cs [hunk 11]
```diff
@@ -98,7 +98,7 @@ internal PowerShellManager CheckoutIdleWorker(StreamingMessage request, AzFuncti
                 }
             }
 
-            // Register the function with the Runspace before returning the idle PowerShellManager.
+            // Register the function with the Runspace PowerShellManager.
```
Comment (daxian-dbw) [inline src/PowerShell/PowerShellManagerPool.cs]: I think the original comment is more clear. Can you please revert this change?

File: src/RequestProcessor.cs (hunk 6 above)
Comment (daxian-dbw) [reply src/RequestProcessor.cs]: Also, a similar issue -- we have to create a `Runspace` and evaluate `profile.ps1` in `FunctionLoad` phase, so that we get a hard failure if `profile.ps1` fails to run.

File: src/RequestProcessor.cs [hunk 12]
```diff
@@ -209,6 +211,28 @@ internal StreamingMessage ProcessInvocationRequest(StreamingMessage request)
...
                 psManager = _powershellPool.CheckoutIdleWorker(request, functionInfo);
 
+                if (_dependencyDownloadTask != null
+                    && ((_dependencyDownloadTask.Status != TaskStatus.Canceled)
+                    || _dependencyDownloadTask.Status != TaskStatus.Faulted
+                    || _dependencyDownloadTask.Status != TaskStatus.RanToCompletion))
+                {
+                    psManager.Logger.Log(LogLevel.Information, PowerShellWorkerStrings.DependencyDownloadInProgress, null, true);
+                    psManager.Logger.Log(LogLevel.Information, PowerShellWorkerStrings.DependencyDownloadInProgress, null);
+                    _dependencyDownloadTask.Wait();
```
Comment (pragnagopa) [inline src/RequestProcessor.cs]: Use async await instead.

**Comment by pragnagopa** [comment]
@daxian-dbw - I will work with @fabiocav to enable PR builds.

File: src/RequestProcessor.cs (hunk 1 above)
Comment (pragnagopa) [reply src/RequestProcessor.cs]: I think better approach would be to return a Rpc messag and block until dependencies are downloaded as invocation will not succeed before download completes.

File: src/RequestProcessor.cs (hunk 6 above)
Comment (pragnagopa) [reply src/RequestProcessor.cs]: Yes. FunctioLoad failure is preserved and is replayed on function invocation. Scope of an invocation request failure is only for that invocation

File: src/RequestProcessor.cs (hunk 1 above)
Comment (pragnagopa) [reply src/RequestProcessor.cs]: We should add another state to the [StatusResult](https://github.com/Azure/azure-functions-language-worker-protobuf/blob/dev/src/proto/FunctionRpc.proto#L112) `Loading`  to allow host to handle intermediate state where function is still loading but not ready for invocation

Comment (daxian-dbw) [reply src/RequestProcessor.cs]: That sounds like a much better long term solution

File: src/RequestProcessor.cs (hunk 6 above)
Comment (SatishRanjan) [reply src/RequestProcessor.cs]: If DependencyError download error occurs then for all invocation will receive this error and function execution wouldn't proceed

Comment (daxian-dbw) [reply src/RequestProcessor.cs]: You are basically moving the logic that's supposed to be handled by the host to the worker.

See my comment https://github.com/Azure/azure-functions-powershell-worker/pull/193#discussion_r277522973
I believe `profile.ps1` needs to be evaluated during function load phase, because a bad `profile.ps1` is a hard failure. But `profile.ps1` cannot be invoked until after the dependency download.

File: src/RequestProcessor.cs (hunk 1 above)
Comment (SatishRanjan) [reply src/RequestProcessor.cs]: The PM requirement is to show the message on the azure portal when a function is invoked through portal while managed dependency download is in-progress

File: src/resources/PowerShellWorkerStrings.resx (hunk 4 above)
Comment (SatishRanjan) [reply src/resources/PowerShellWorkerStrings.resx]: These spaces are auto formatted

File: src/RequestProcessor.cs (hunk 5 above)
Comment (SatishRanjan) [reply src/RequestProcessor.cs]: System log is to appear on CLI console

File: src/RequestProcessor.cs (hunk 6 above)
Comment (SatishRanjan) [reply src/RequestProcessor.cs]: Since dependency download happens on a separate instance of the PowerShell and doesn't use the runspace of powershell pool  hence I've added back the __powershellPool.Initialize(request.RequestId); as part of FunctionLoad

File: src/PowerShell/PowerShellManagerPool.cs (hunk 10 above)
Comment (SatishRanjan) [reply src/PowerShell/PowerShellManagerPool.cs]: Removed

**Comment by pragnagopa** [comment]
Will enable builds ASAP. Will update this PR soon.

File: src/DependencyManagement/DependencyManager.cs [hunk 13]
```diff
@@ -51,15 +59,83 @@ internal class DependencyManager
...
+        internal void ProcessDependencyDownload(MessagingStream msgStream, StreamingMessage request)
+        {
+            if (request.FunctionLoadRequest.ManagedDependencyEnabled)
+            {
+                var rpcLogger = new RpcLogger(msgStream);
+                rpcLogger.SetContext(request.RequestId, null);
+                if (!_shouldUpdateFunctionAppDependencies)
+                {
+                    // The function app already has the latest dependencies installed.
+                    rpcLogger.Log(LogLevel.Information, PowerShellWorkerStrings.LatestFunctionAppDependenciesAlreadyInstalled, isUserLog: true);
```
Comment (daxian-dbw) [inline src/DependencyManagement/DependencyManager.cs]: This message is not accurate. It's possible the `requirements.ps1` only contains an empty hashable -- that's the way for a user to opt out the managed dependency if they don't depend on Az module at all.
Please take a look at how log messages are handled in `InstallFunctionAppDependencies` when `if (Dependencies.Count == 0)` and `if (!_shouldUpdateFunctionAppDependencies)`, and it should be handled the same way.

File: src/DependencyManagement/DependencyManager.cs [hunk 14]
```diff
@@ -51,15 +59,83 @@ internal class DependencyManager
         // Managed Dependencies folder name.
         private const string ManagedDependenciesFolderName = "ManagedDependencies";
 
+        private volatile Exception _dependencyError;
+
         // This flag is used to figure out if we need to install/reinstall all the function app dependencies.
         // If we do, we use it to clean up the module destination path.
         private bool _shouldUpdateFunctionAppDependencies;
 
+        private volatile Task _dependencyDownloadTask;
```
Comment (daxian-dbw) [inline src/DependencyManagement/DependencyManager.cs]: `volatile` is not needed for both `_dependencyError` and `_dependencyDownloadTask`. They are not being modified on two threads at the same time.

File: src/DependencyManagement/DependencyManager.cs [hunk 15]
```diff
@@ -51,15 +59,83 @@ internal class DependencyManager
...
+        {
+            if (_dependencyDownloadTask != null)
+            {
+                _dependencyDownloadTask.Wait();
+                _dependencyDownloadTask = null;
+            }
+        }
+
+        private void ProcessDependencies(
+            MessagingStream msgStream,
```
Comment (daxian-dbw) [inline src/DependencyManagement/DependencyManager.cs]: Parameter not used.

File: src/DependencyManagement/DependencyManager.cs [hunk 16]
```diff
@@ -51,15 +59,83 @@ internal class DependencyManager
...
+            StreamingMessage request,
```
Comment (daxian-dbw) [inline src/DependencyManagement/DependencyManager.cs]: Parameter not used.

File: src/DependencyManagement/DependencyManager.cs [hunk 17]
```diff
@@ -28,6 +31,11 @@ internal class DependencyManager
         // This is the location where the dependent modules will be installed.
         internal static string DependenciesPath { get; private set; }
 
+        internal Exception DependencyError { get => _dependencyError; }
+
+        //The dependency download task
+        internal Task DependencyDownloadTask { get => _dependencyDownloadTask; }
```
Comment (daxian-dbw) [inline src/DependencyManagement/DependencyManager.cs]:
```suggestion
        internal Task DependencyDownloadTask => _dependencyDownloadTask;
```

File: src/DependencyManagement/DependencyManager.cs [hunk 18]
```diff
@@ -28,6 +31,11 @@ internal class DependencyManager
         // This is the location where the dependent modules will be installed.
         internal static string DependenciesPath { get; private set; }
 
+        internal Exception DependencyError { get => _dependencyError; }
```
Comment (daxian-dbw) [inline src/DependencyManagement/DependencyManager.cs]:
```suggestion
        internal Exception DependencyError => _dependencyError;
```

**Comment by fabiocav** [comment]
Reopening PR to test CI integration

File: src/RequestProcessor.cs [hunk 19]
```diff
@@ -206,9 +205,33 @@ internal StreamingMessage ProcessInvocationRequest(StreamingMessage request)
...
             {
+                if (_dependencyManager.DependencyDownloadTask != null
+                    && ((_dependencyManager.DependencyDownloadTask.Status != TaskStatus.Canceled)
+                    || _dependencyManager.DependencyDownloadTask.Status != TaskStatus.Faulted
+                    || _dependencyManager.DependencyDownloadTask.Status != TaskStatus.RanToCompletion))
+                {
+                    var rpcLogger = new RpcLogger(_msgStream);
+                    rpcLogger.SetContext(request.RequestId, request.InvocationRequest?.InvocationId);
+                    rpcLogger.Log(LogLevel.Information, PowerShellWorkerStrings.DependencyDownloadInProgress, null, true);
+                    rpcLogger.Log(LogLevel.Information, PowerShellWorkerStrings.DependencyDownloadInProgress, null);
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/RequestProcessor.cs]: #closed.

File: src/RequestProcessor.cs [hunk 20]
```diff
@@ -206,9 +205,33 @@ internal StreamingMessage ProcessInvocationRequest(StreamingMessage request)
...
+                    _dependencyManager.DependencyDownloadTask.Wait();
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/RequestProcessor.cs]: #closed.

File: src/DependencyManagement/DependencyManager.cs [hunk 21]
```diff
@@ -60,6 +74,74 @@ internal DependencyManager()
...
+        internal void ProcessDependencyDownload(MessagingStream msgStream, StreamingMessage request)
+        {
+            if (request.FunctionLoadRequest.ManagedDependencyEnabled)
+            {
+                var rpcLogger = new RpcLogger(msgStream);
+                rpcLogger.SetContext(request.RequestId, null);
+                if (Dependencies.Count == 0)
+                {
+                    // If there are no dependencies to install, log and return.
+                    rpcLogger.Log(LogLevel.Trace, PowerShellWorkerStrings.FunctionAppDoesNotHaveDependentModulesToInstall, isUserLog: true);
```
Comment (daxian-dbw) [inline src/DependencyManagement/DependencyManager.cs]: Could it be possible that the portal looks like freezing currently without any message only because we are using the `Trace` level logging? Maybe if you change the current logging to `Information` level and the message will show up even without this change?

File: src/RequestProcessor.cs [hunk 22]
```diff
@@ -299,28 +318,14 @@ internal StreamingMessage ProcessFunctionEnvironmentReloadRequest(StreamingMessa
         private void InitializeForFunctionApp(StreamingMessage request, StreamingMessage response)
```
[the remark this answers is not in the document]
Comment (SatishRanjan) [reply src/RequestProcessor.cs]: I'll perform the protobuf cleanup in the next PR

File: src/DependencyManagement/DependencyManager.cs (hunk 21 above)
Comment (SatishRanjan) [reply src/DependencyManagement/DependencyManager.cs]: I tried changing the trace level but it doesn't show up on azure portal unless it's exception.

File: src/RequestProcessor.cs (hunk 22 above)
[the remark this answers is not in the document]
Comment (SatishRanjan) [reply src/RequestProcessor.cs]: Cleaned up the StreamingMessage response parameter

File: src/RequestProcessor.cs [hunk 23]
```diff
@@ -225,7 +247,6 @@ internal StreamingMessage ProcessInvocationRequest(StreamingMessage request)
             catch (Exception e)
             {
                 _powershellPool.ReclaimUsedWorker(psManager);
-
```
[the remark this answers is not in the document]
Comment (SatishRanjan) [reply src/RequestProcessor.cs]: Added the new line

File: src/RequestProcessor.cs [hunk 24]
```diff
@@ -234,7 +255,6 @@ internal StreamingMessage ProcessInvocationRequest(StreamingMessage request)
                 response.InvocationResponse.InvocationId = request.InvocationRequest.InvocationId;
                 status.Status = StatusResult.Types.Status.Failure;
                 status.Exception = e.ToRpcException();
-
```
[the remark this answers is not in the document]
Comment (SatishRanjan) [reply src/RequestProcessor.cs]: added new line
